Common Issues content from Astrocloud PR (#350)
astronomer/docs: astro/test-and-troubleshoot-locally.md

astro dev kill
```

> This command forces your running containers to stop and deletes all data associated with your local Postgres metadata database, including Airflow Connections, logs, and task history.

Reviewer: ```suggestion
This command forces your running containers to stop and deletes all data associated with your local Postgres metadata database, including Airflow connections, logs, and task history.
```

## Troubleshoot common issues

> Use the information provided here to resolve common issues with Astronomer projects in local environments.

Reviewer: ```suggestion
Use the information provided here to resolve common issues with running an Astro project in a local environment.
```

### New DAGs aren't visible in the Airflow UI

By default, the Airflow scheduler scans the DAGs directory for new files every 300 seconds (5 minutes). For this reason, it might take a few minutes for DAGs to appear in the Airflow UI.

To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. This setting is dependent on the amount of CPU that is allocated to the scheduler. The greater the CPU allocation, the faster the scheduler can parse your DAGs.

### DAGs are running slowly

If your Astro project contains many DAGs or tasks, then you might experience performance issues in your local Airflow environment.

> To improve the performance of your environment, you can adjust CPU and memory resource allocation in your Docker resources configuration. Be aware that increasing Docker resource allocation might decrease the overall performance of your computer.
> 
> You can also modify Airflow-level settings to improve the performance of your projects. For more information, see  [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers) guide from Astronomer.
> 
> If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to an Astro Deployment that's dedicated to testing.

Reviewer: ```suggestion
To improve the performance of your environment, you can:

 - Adjust CPU and memory resource allocation in your Docker resources configuration. Be aware that increasing Docker resource allocation might decrease the overall performance of your computer.
- Modify Airflow-level environment variables to improve the performance of your local environment, including concurrency and parallelism. For more information, see  [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers) guide from Astronomer.

If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to a Deployment on Astro that's dedicated to testing.
```

### Astro project won't load after `astro dev start`

> If your Astro project fails to run in an Airflow environment more than 5 minutes after running `astro dev start`, it might be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:

Reviewer: ```suggestion
If your Astro project fails to run in an Airflow environment for more than 5 minutes after running `astro dev start`, it might be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:
```

To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. This setting is dependent on the amount of CPU that is allocated to the scheduler. The greater the CPU allocation, the faster the scheduler can parse your DAGs.

### DAGs are running slowly

If your Astro project contains many DAGs or tasks, then you might experience performance issues in your local Airflow environment.

To improve the performance of your environment, you can:

 - Adjust CPU and memory resource allocation in your Docker resources configuration. Be aware that increasing Docker resource allocation might decrease the performance of your computer.
- Modify Airflow-level environment variables to improve the performance of your local environment, including concurrency and parallelism. See  [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers).

If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to a Deployment on Astro that's dedicated to testing.

> 

Reviewer 2: Dynamic DAG authoring patterns are a common culprit here also I believe.

Reviewer 3: @dylanbstorey I attempted to add the information you suggested. Please review and edit as necessary.

Reviewer: @dylanbstorey Good suggestion here. Any written content or examples you'd point to that's more specific? Where can a user go read about "dynamic dag authoring patterns" if they don't quite know what that means, and whether or not it applies to them?

Reviewer 2: > - Make sure you aren't using dynamic DAG authoring patterns

This is a bit strong. How about something like : 

If you're using dynamic DAG authoring patterns you may need to tune them. Please look to [this](https://www.astronomer.io/guides/dynamically-generating-dags/) guide for more information.

Reviewer 2: @paolaperaza  - https://www.astronomer.io/guides/dynamically-generating-dags/

Reviewer 4: Maybe this @paolaperaza? https://www.astronomer.io/guides/dynamically-generating-dags/

Reviewer 3: @ben-astro and @dylanbstorey I have adjusted the text and added a doc link. Please note that I need to keep the bulleted list parallel, so I need to use the same type of language for each item.

Reviewer 5: I don't think we should recommend they change their behavior, but rather let them know that their behavior might not work well with local development.

### Astro project won't load after `astro dev start`

If your Astro project fails to run in an Airflow environment for more than 5 minutes after running `astro dev start`, it might be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:

1. After running `astro dev start`, retrieve a list of running containers by running `astro dev ps`.
2. If the webserver and scheduler containers exist but are unhealthy, check their logs by running:

    ```sh
    $ astro dev logs --webserver
    $ astro dev logs --scheduler
    ```

These logs should help you understand why your webserver or scheduler is unhealthy. Possible reasons why these containers might be unhealthy include:

> - Not enough Docker resources.

Reviewer 2: I'd probably add something like : `consider running docker system prune --volumes` here. I've seen a number of users running out of disk space in the docker daemon and the error message doesn't do a great job of telling you what the underlying cause is.

- Broken Postgres database.

> - DAG code errors.

Reviewer 2: I'm not sure an improperly formatted  DAG can keep the services from starting. A plugin might be able to though.

Reviewer 3: @dylanbstorey I removed _DAG code errors_.

### DAGs are running slowly

If your Astro project contains many DAGs or tasks, then you might experience performance issues in your local Airflow environment.

To improve the performance of your environment, you can:

 - Adjust CPU and memory resource allocation in your Docker resources configuration. Be aware that increasing Docker resource allocation might decrease the performance of your computer.
- Modify Airflow-level environment variables to improve the performance of your local environment, including concurrency and parallelism. See  [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers).
- Adjust your dynamic DAG authoring patterns. See [Dynamically Generating DAGs in Airflow](https://www.astronomer.io/guides/airflow-scaling-workers).

If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to a Deployment on Astro that's dedicated to testing.

### Astro project won't load after `astro dev start`

> If your Astro project fails to run in an Airflow environment for more than 5 minutes after running `astro dev start`, it might be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:

Reviewer: ```suggestion
If you're running the Astro CLI on a Mac computer that's built with the Apple M1 chip, your Astro project might take more than 5 mins to start after running `astro dev start`. This is a current limitation of Astro Runtime and the Astro CLI that's expected to be addressed soon.

If your project won't load, it may also be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:
```

Added a statement on Apple M1 issues per @ben-astro's recommendation. We are fixing this issue super soon, but it's a very common reason for why `astro dev start` takes forever, so I want to make sure to include it. Once the feature comes out, we can come back to this doc.

```

This command forces your running containers to stop and deletes all data associated with your local Postgres metadata database, including Airflow connections, logs, and task history.

## Troubleshoot common issues

Use the information provided here to resolve common issues with running an Astro project in a local environment.

### New DAGs aren't visible in the Airflow UI

By default, the Airflow scheduler scans the DAGs directory for new files every 300 seconds (5 minutes). For this reason, it might take a few minutes for DAGs to appear in the Airflow UI.

> To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. This setting is dependent on the amount of CPU that is allocated to the scheduler. The greater the CPU allocation, the faster the scheduler can parse your DAGs.

Reviewer: ```suggestion
To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. See [Set environment variables on Astro](environment-variables.md). This setting is dependent on the amount of CPU that is allocated to the scheduler. The greater the CPU allocation, the faster the scheduler can parse your DAGs.
```

> Make sure the `dag_id` isn’t duplicated. When two DAGs use the same `dag_id`, the newest DAG won't appear in the Airflow UI and you won't receive an error message.

Reviewer: ```suggestion
If your DAGs aren't visible in the Airflow UI, first make sure that `dag_id` isn’t duplicated across the DAGs in your local environment. When two DAGs use the same `dag_id`, the newest DAG won't appear in the Airflow UI and you won't receive an error message.
```

@kwcanuck I think this is an easier "fix" and overall best practice -- can you bump up this section to the top, and then have the rest be, "If you have no duplicate dag ids, you can configure your local airflow environment to parse your dags more often..." (or something like that to introduce it)

Reviewer 3: @paolaperaza I don't think the text _If your DAGs aren't visible in the Airflow UI, first make sure that..._ is necessary as this information below the heading _New DAGs aren't visible in the Airflow UI_. What you are suggesting is a duplication of information. 

I don't think a_cross the DAGs in your local environment_ adds any value. It's clear that the issue could be related to a duplicated dag-id, no matter where it appears.

- Modify Airflow-level environment variables to improve the performance of your local environment, including concurrency and parallelism. See  [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers).
- Make sure you aren't using dynamic DAG authoring patterns.

If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to a Deployment on Astro that's dedicated to testing.

### Astro project won't load after `astro dev start`

If your Astro project fails to run in an Airflow environment for more than 5 minutes after running `astro dev start`, it might be because your webserver or scheduler is unhealthy. In this case, you might need to debug your containers. To do so:

1. After running `astro dev start`, retrieve a list of running containers by running `astro dev ps`.
2. If the webserver and scheduler containers exist but are unhealthy, check their logs by running:

    ```sh
    $ astro dev logs --webserver

>     $ astro dev logs --scheduler

Reviewer: @dylanbstorey Would you say `astro dev logs` or `docker exec -it <scheduler>` is more useful here? If the latter, what would a user look for in those containers?

Reviewer 2: astro dev logs. 

Executing into the container and poking around might help them investigate where a file wound up or why a path doesn't look like they think it should, but it won't help them solve the issue long term.

```

This command forces your running containers to stop and deletes all data associated with your local Postgres metadata database, including Airflow Connections, logs, and task history.

> ## Troubleshooting Common Issues

Reviewer 3: ```suggestion
## Troubleshoot Common Issues
```

> You might experience one of the following common issues when testing your Astronomer project in a local environment. This section contains information about how to approach each of these issues.

Reviewer 3: Question: Do we want to use _common_? As a user, my response to this would be:  _If they are common, why doesn't Astronomer fix them?_

> You might experience one of the following common issues when testing your Astronomer project in a local environment. This section contains information about how to approach each of these issues.

Reviewer 3: ```suggestion
Use the information provided here to resolve common issues with Astronomer projects in a local environments.
```

### New DAGs aren't visible in the Airflow UI

> By default, the Airflow scheduler scans the DAGs directory for new files every 300 seconds (5 minutes). For this reason, it might take a minute or two to see new DAGs appear in the Airflow UI.

Reviewer 3: ```suggestion
By default, the Airflow scheduler scans the DAGs directory for new files every 300 seconds (5 minutes). For this reason, it might take a few minutes for DAGs to appear in the Airflow UI.
```

> To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. Note that this setting is dependent on how much CPU is available to the scheduler: the more available CPU, the faster the scheduler can parse your DAGs.

Reviewer 3: ```suggestion
To have the scheduler check for DAGs more frequently, you can set the `AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL` environment variable to less than 300 seconds. This setting is dependent on the amount of CPU that is allocated to the scheduler. The greater the CPU allocation, the faster the scheduler can parse your DAGs.
```

### DAGs are running slowly

If your Astro project contains many DAGs or tasks, then you might experience performance issues in your local Airflow environment.

> To improve your environment's performance, you can scale both CPUs and memory in your Docker resources configuration. Note that increasing Docker's resource usage might decrease the overall performance of your machine.

Reviewer 3: ```suggestion
To improve the performance of your environment, you can adjust CPU and memory resource allocation in your Docker resources configuration. Increasing Docker resource allocation might decrease the overall performance of your computer.
```

You can also modify Airflow-level settings to improve your project's performance. For more information, read Astronomer's guide to [Scaling out Airflow](https://www.astronomer.io/guides/airflow-scaling-workers).

> If your DAGs are still running slowly and you cannot scale Docker or Airflow any further, then we recommend pushing your project to an Astro Deployment that's dedicated to testing.

Reviewer 3: ```suggestion
If your DAGs continue to run slowly and you can't scale Docker or Airflow any further, Astronomer recommends pushing your project to an Astro Deployment that's dedicated to testing.
```
